[IDEA plugin] Create Avail project wizard

raulraja opened the pull request:
Fixes #307.

Creates a new version of the create project dialog similar to what was available in Anvil.
Removes fields from the project dialog that were related to paths and projects since those are handled by IDEA.

This code still does not handle more things we need to implement, such as adding the `roots` as a source element in a module where the `avail` language can be activated and a VM can run compiling and parsing the avail sources.

I think we should consider when creating a new empty project to automatically add `gradle` or whatever build tool we think most people will use when building avail code. Currently, the created project is not useful. It does not include a Hello World or something that could get people started.

https://github.com/AvailLang/Avail/assets/456796/d01c074d-5040-4d12-b396-16db0f611ff5

raulraja committed:
Draft for Create Avail Project Wizard

raulraja committed:
Code refactor to separate views from project creation logic

raulraja committed:
Better styles for swing component

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailModuleWizardStep.kt:
```diff
@@ -31,24 +31,50 @@
+		}
 	}
 
+	private val Module.config get() =
```
Add documentation comment

richATavail replied:
@raulraja Ok, I apologize, but I'm going to be a bit particular here. I'd update the comment to take this form:

```kotlin
/**
 * Given a Module and the [createProjectPanel] fields, create the
 * [AvailProjectTemplate.Config] for the project to be created.
 */
```

Putting the `AvailProjectTemplate.Config` in braces guarantees that on a refactor-rename, this will get updated to the new name in the comment. Without it, this mention wouldn't be updated. We've learned the hard way that the tools don't always update these mentions in comments on a rename.

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailModuleWizardStep.kt:
```diff
@@ -31,24 +31,50 @@
+			)
+		}
+
+	private val createProjectPanel: CreateAvailProjectPanel =
```
Add documentation comment

richATavail replied:
@raulraja We work to keep code lines to run no more than 80 characters. This enables people with differing screen sizes to maximize the number of code editor panes. 

```kotlin
/**
 * Represents the panel used to create a new project.
 *
 * @property createProjectPanel 
 *   The instance of the CreateAvailProjectPanel class used to display the 
 *   create project form.
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailModuleWizardStep.kt:
```diff
@@ -31,24 +31,50 @@
+				rootsDir = panel.rootsDirField.input.ifEmpty { "roots" },
+				rootName = panel.rootNameField.input.ifEmpty { name },
+				importStyles = panel.importStyles.isSelected,
+				importTemplates = panel.importTemplates.isSelected,
```
Should we handle template expansions natively via IntelliJ's tools? Templates as they exist are functionally specific to an Anvil implementation at this point. The template itself is generic enough probably use across IDEs; maybe we can create client specific imports for carrying template expansions across IDEs?

raulraja replied:
The common practice in this kind of plugin seems to be having the initial project layout copied from the `resources` folder into the `target` module folder. When they do this, they support Velocity templates, which help variable replacements in case any parts of the project layout are dynamic. The velocity templates are then evaluated, and the final layout is created.

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+ */
+object AvailProjectTemplate {
+
+  data class Config(
```
Add documentation comments for the data class and the properties.

richATavail replied:
Our comment style is to place the description portion of each `@property` on the next line two-spaces in. We also want the comment lines to be limited to 80 characters:

```kotlin
/**
 * Represents a configuration for an Avail software project.
 *
 * @property projectLocation
 *   The location of the project.
 * @property fileName
 *   The name of the configuration file.
 * @property rootsDir
 *   The directory where the project's root files are located.
 * @property rootName
 *   The name of the project's root file.
 * @property importStyles
 *   A flag indicating whether to import styles.
 * @property importTemplates
 *   A flag indicating whether to import templates.
 * @property libraryName
 *   The name of the library used in the project (optional).
 * @property selectedLibrary
 *   The selected library file (optional).
 * @property environmentSettings
 *  The global environment settings for the project.
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/CreateAvailProjectPanel.kt:
```diff
@@ -0,0 +1,188 @@
+
+		// The library picker
+		c.gridy = 3
+		add(libraryPicker, c)
```
Maybe add a label or header above indicating what this is? Or maybe hover text?
![image](https://github.com/AvailLang/Avail/assets/13374407/0fb8ac8a-c4db-4231-9528-815d6eb2389d)

raulraja replied:
With the latest changes in https://github.com/AvailLang/Avail/pull/310/commits/856dba8ff86587e9045a08d0ce08bb7928811895 it should look now like:
<img width="783" alt="Screenshot 2023-10-16 at 20 21 24" src="https://github.com/AvailLang/Avail/assets/456796/9f81dbe7-329f-43f9-8253-6a32ac5d0dd7">

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+    }
+  }
+
+  private fun AvailProjectV1.updateBackingProjectFile(
```
add a documentation comment.

raulraja replied:
Done in https://github.com/AvailLang/Avail/pull/310/commits/377e9c11de82c578bba8149a0c61fe573d53655a

richATavail replied:
Wrap the `@param` comment descriptions. Consider adding documentation links in the comment to enable better IDE comment navigation. Something like:

```kotlin
/**
 * Updates the backing project file with the current [AvailProjectV1]
 * instance.
 *
 * @param projectFilePath
 *   The file path of the backing project file.
 * @param config
 *   The [configuration][Config] settings.
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+    }
+  }
+
+  private fun AvailProjectV1.configureAvailStdLib(
```
add a documentation comment.

raulraja replied:
Done in https://github.com/AvailLang/Avail/pull/310/commits/377e9c11de82c578bba8149a0c61fe573d53655a

richATavail replied:
Update the comment to follow the formatting standards described above:

```kotlin
/**
 * Configures the Avail standard library for the given Avail project.
 *
 * @param config 
 *   The [configuration][Config] settings for the Avail project.
 * @param lib 
 *   The file representing the Avail standard library.
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+    addRoot(stdLib)
+  }
+
+  private fun importTemplates(
```
add a documentation comment.

raulraja replied:
Done in https://github.com/AvailLang/Avail/pull/310/commits/377e9c11de82c578bba8149a0c61fe573d53655a

richATavail replied:
Per my previous comments on comments, update this to something like:

```kotlin
/**
 * Imports the templates from the given [AvailArtifactJar] and returns them as 
 * a [TemplateGroup].
 *
 * @param config 
 *   The [configuration][Config] object that specifies whether to import 
 *   templates.
 * @param jar 
 *   The [AvailArtifactJar] from which to import the templates.
 * @return 
 *   The imported templates as a [TemplateGroup].
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+    TemplateGroup()
+  }
+
+  private fun importStyles(
```
add a documentation comment.

raulraja replied:
Done in https://github.com/AvailLang/Avail/pull/310/commits/377e9c11de82c578bba8149a0c61fe573d53655a

richATavail replied:
Update the comment to follow the commenting style. Something like:

```kotlin
/**
 * Imports the styles for the given Avail artifact jar.
 *
 * @param config 
 *   The [configuration][Config] object containing import options.
 * @param jar 
 *   The Avail artifact jar from which to import styles.
 * @return 
 *   The imported [styling group][StylingGroup] if importStyles is enabled in 
 *   the config object, otherwise an empty styling group.
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+    StylingGroup()
+  }
+
+  private fun availProjectRoot(
```
add a documentation comment.

richATavail replied:
Update this comment to match the previously stated standards. Something like:

```kotlin
/**
 * Creates a new [AvailProjectRoot] object initialized with the provided 
 * parameters.
 *
 * @param rootConfigDir 
 *   The root configuration directory.
 * @param config 
 *   The project [configuration][Config].
 * @param rootName 
 *   The [root name][AvailProjectRoot.name].
 * @param rootLocationDir 
 *   The [root location directory][AvailProjectRoot.location].
 * @return 
 *   The newly created []AvailProjectRoot] object.
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+ * @author Raul Raja
+ *
+ */
+object AvailProjectTemplate {
```
The Avail project style puts the opening curly brace for class/interfaces/objects on the next line:
```kotlin
object AvailProjectTemplate
{
```

richATavail replied:
@raulraja move the curly brace, `{` to the next line.

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+    projectFilePath: String,
+    config: Config
+  ) {
+    if (projectFilePath.isNotEmpty()) {
```
The Avail style of code puts the curly brace for it blocks on the next line
```kotlin
if (projectFilePath.isNotEmpty)
{
```

richATavail replied:
Please place the open curly brace, `{` on the next line.

This can be configured just for the project in Settings:
![image](https://github.com/AvailLang/Avail/assets/13374407/8add00a0-f0dc-4829-a710-eda362c932fc)

I tried to export a settings file that you could import for the project to make this easy, but IntelliJ seems to have broken the export to no longer include some of the settings. Follow up with @toddATavail to get more help on setting up the environment in a way that will make formatting easier. This will also automatically have the project use tabs instead of spaces for indentation.

It may be worthwhile for us to investigate some kind of CI/CD formatting tools that auto-format the project on check-in to match the style requirements for Avail. We haven't done this previously as the team pretty much follows the same format in  both Avail programming and their own projects/work outside of Avail. Getting something like this going will reduce the onus on other volunteers working on the Avail code base.

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+
+    val rootManifest = jar.manifest.roots[AVAIL_STDLIB_ROOT_NAME]
+    var sg =
+      importStyles(config, jar)
```
this should fit on the previous line with the variable declaration

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+    var sg =
+      importStyles(config, jar)
+    var tg =
+      importTemplates(config, jar)
```
this should fit on the previous line with the variable declaration

raulraja replied:
Addressed in https://github.com/AvailLang/Avail/pull/310/commits/856dba8ff86587e9045a08d0ce08bb7928811895

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+      ?: TemplateGroup()
+  } else {
+    TemplateGroup()
+  }
```
Adjust curly braces to the next line:
```kotlin
if (config.importTemplates)
{
	...
}
else
{
	...
}
```

raulraja replied:
See https://github.com/AvailLang/Avail/pull/310/commits/856dba8ff86587e9045a08d0ce08bb7928811895

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+      ?: StylingGroup()
+  } else {
+    StylingGroup()
+  }
```
Adjust the positions of the `if-else` blocks.

raulraja replied:
Done in https://github.com/AvailLang/Avail/pull/310/commits/856dba8ff86587e9045a08d0ce08bb7928811895

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+    "$rootsDir/$rootName"
+  } else {
+    rootName
+  }
```
The first line exceeds 80 characters. Format code:

```kotlin
private fun rootsLocationDir(rootsDir: String, rootName: String) =
	if (rootsDir.isNotEmpty())
	{
		"$rootsDir/$rootName"
	}
	else
	{
		rootName
	}
```

Because each branch block presents a single short expression, I think it is also stylistically acceptable in Avail to do this (_at least I haven't been told not to_ 😅). In my opinion it is nice short and clean and isn't confusing. The moment the branch blocks become multi-line or multi-statement, I tend to add curly braces.
```kotlin
private fun rootsLocationDir(rootsDir: String, rootName: String) =
	if (rootsDir.isNotEmpty()) "$rootsDir/$rootName"
	else rootName
```

raulraja replied:
I chose your shorter style for the last one since it had fewer code in the branches. https://github.com/AvailLang/Avail/pull/310/commits/856dba8ff86587e9045a08d0ce08bb7928811895

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,185 @@
+
+  /**
+   * Create the Avail project.
+   */
```
add an `@param config` comment explaining what the `Config` is for.

raulraja replied:
Done in https://github.com/AvailLang/Avail/pull/310/commits/856dba8ff86587e9045a08d0ce08bb7928811895

richATavail reviewed (changes requested):
Its coming along nicely.  A couple of points:

1. The Avail project code base uses tabs instead of spaces for new line indentation. @toddATavail has an IntelliJ settings file he can share to help with formatting.
2. We generally follow [C#'s coding conventions](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions) when it comes to brace placement in code. 
3. We work to keep code lines to a maximum of 80 characters to enable multiple code panes open side-by-side. I only observed one line that ran long.

Additionally, what are your thoughts on starting [localization](https://plugins.jetbrains.com/docs/intellij/localization-guide.html) early? We clearly were not doing that in Anvil, but maybe we should consider it given the Avail has powerful features to easily create standard libraries in other spoken/written languages; it would be kind of weird to be able to program in Spanish or Korean but need to setup your project in English. Just a thought.

raulraja committed:
Addresses PR comments

raulraja committed:
Add missing method docs

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailModuleWizardStep.kt:
```diff
@@ -31,24 +31,61 @@
  * @author Richard Arriaga
  */
-class AvailModuleWizardStep: ModuleWizardStep()
+class AvailModuleWizardStep(builder: AvailModuleBuilder) : ModuleWizardStep()
```
The Avail style dictates that adding a non-trivial constructor requires a comment be added for constructor if the constructor inputs are not class fields with `@property` documentation comments

```kotlin
/**
 * The [ModuleWizardStep] used to set up a new Avail project.
 *
 * @author Richard Arriaga
 * 
 * @constructor
 * Construct an [AvailModuleWizardStep].
 * 
 * @param builder
 *   The [AvailModuleBuilder] used to construct the Avail [Module].
 */
```

The `[Module]` reference requires the addition of the following import statement to create the link to the documentation of `Module`:

```kotlin
import com.intellij.openapi.module.Module
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+   * Creates a new Avail project.
+   *
+   * @param config The configuration for the project.
+   */
```
We want to wrap the comment after `@param config`. Also, adding document links to types in the comments helps when navigating the comments in the IDE.

```kotlin
/**
 * Creates a new [AvailProject].
 *
 * @param config
 *   The [Config] used to configure the project.
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+   * @param config The configuration settings.
+   */
+  private fun AvailProjectV1.updateBackingProjectFile(
+    projectFilePath: String,
+    config: Config
+  )
+  {
```
Move `)` and `{` to the same line. We used to be very particular about moving the `{` to its own line while keeping `)` on the line of the last argument, now we've taken to combining them on the next line if there 

1. There is more than one parameter
2. There is only one parameter but the whole function signature doesn't fit on one line
3. There is no return type

Here are examples:

```kotlin
fun fooAndSuch (
	qqqq: String,
	zzz: Double
){

////////////

fun fooAndSuch (
	qqqq: String,
	zzz: Double
): Baz
{

///////////

fun fooAndSuch (qqqq: String): Baz
{
///////////////////////

fun superReallyExtremlyVeryLongSignatureThatJustKeepsOnGoing(
	qqq: Long
): Baz
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+  private fun AvailProjectV1.configureAvailStdLib(
+    config: Config,
+    lib: File
+  )
```
move the curly brace, `{`, up to the same line as `)`:

```kotlin
){
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+    var sg = importStyles(config, jar)
+    var tg = importTemplates(config, jar)
+    val extensions = mutableListOf<String>()
+    var description = ""
+    rootManifest?.let {
+      sg = it.styles
+      tg = it.templates
+      extensions.addAll(it.availModuleExtensions)
+      description = it.description
+    }
```
I know I wrote this code and you are cloning it, but feel free to do a cooler job than I did. I'm often pretty critical of the code I write as I like to keep getting better and writing cleaner code. So feel free when cloning something I did to improve upon it. I'm not sensitive about these things and I'm very comfortable in recognizing when code is done better than I originally did. This comment isn't about changing this code, just an acknowledgement that if there is a better, more Kotlin way to do something, than feel free to make those changes. 😀

With this for example, just a cursory sweep, I could've made this be a little neater (_probably could improve it some more than below_):

```kotlin
val description = rootManifest?.let {
	sg = it.styles
	tg = it.templates
	extensions.addAll(it.availModuleExtensions)
	it.description
} ?: ""
addRoot(
	AvailProjectRoot(
		libConfigDir,
		projLocation,
		libName,
		AvailLibraries(
		"org/availlang/${lib.name}",
			Scheme.JAR,
			AVAIL_STDLIB_ROOT_NAME),
		LocalSettings(
			AvailEnvironment.projectRootConfigPath(
				fileName,
				libName,
				projLocation)),
		sg,
		tg,
		extensions
	).apply {
		this.description = description
		saveLocalSettingsToDisk()
		saveTemplatesToDisk()
		saveStylesToDisk()
	})
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+    jar: AvailArtifactJar
+  ) = if (config.importTemplates)
+  {
+    jar.manifest.templatesFor(AVAIL_STDLIB_ROOT_NAME)
+      ?: TemplateGroup()
```
I don't think Avail has a preference one way or another, but this could fit on one line (_less than 80 characters_):
```kotlin
jar.manifest.templatesFor(AVAIL_STDLIB_ROOT_NAME) ?: TemplateGroup()
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+  {
+    jar.manifest.templatesFor(AVAIL_STDLIB_ROOT_NAME)
+      ?: TemplateGroup()
+  } else
```
Move `else` to the next line:

```kotlin
}
else
{
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+    jar: AvailArtifactJar
+  ) = if (config.importStyles)
+  {
+    jar.manifest.stylesFor(AVAIL_STDLIB_ROOT_NAME)
+      ?: StylingGroup()
```
This too could fit on one line (_less than 80 characters_)

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+  {
+    jar.manifest.stylesFor(AVAIL_STDLIB_ROOT_NAME)
+      ?: StylingGroup()
+  } else
```
Move `else` to its own line

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+    TemplateGroup()
+  )
+
+  /**
+   * Concatenates the given root directory and root name to form a new location directory.
+   *
+   * @param rootsDir The root directory string.
+   * @param rootName The root name string.
+   * @return The location directory string formed by concatenating rootsDir and rootName.
+   */
```
Reformat the comment:

```kotlin
/**
 * Concatenates the given root directory and root name to form a new 
 * location directory.
 *
 * @param rootsDir 
 *   The root directory string.
 * @param rootName 
 *   The root name string.
 * @return 
 *   The location directory string formed by concatenating rootsDir and 
 *   rootName.
 */
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailProjectTemplate.kt:
```diff
@@ -0,0 +1,244 @@
+   * @param rootName The root name string.
+   * @return The location directory string formed by concatenating rootsDir and rootName.
+   */
+  private fun rootsLocationDir(rootsDir: String, rootName: String) =
+    if (rootsDir.isNotEmpty()) "$rootsDir/$rootName"
+    else rootName
```
Specify the return type. I know Kotlin can infer it, but a developer scanning the code will need to stop and do a bit more digging to know what the return type is (e.g. _hover over the function name in the IDE_). Maybe reformat it to something like:

```kotlin
private fun rootsLocationDir(
	rootsDir: String,
	rootName: String
): String =
	if (rootsDir.isNotEmpty()) "$rootsDir/$rootName"
	else rootName
```

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/CreateAvailProjectPanel.kt:
```diff
@@ -0,0 +1,185 @@
+ *
+ * @author Raul Raja
+ *
+ * @property config
+ *   The [GlobalEnvironmentSettings] for this machine.
+ * @property onCreate
+ *   The function that accepts the newly created [AvailProject].
+ * @property onCancel
+ *   The function to call if creating a new project is canceled.
```
None of these fields appear to exist in `CreateAvailProjectPanel`; looks like a cloning error from `avail.anvil.manager.CreateProjectPanel`. Please remove these comments.

richATavail reviewed (changes requested):
I've added more requests for changes to bring things stylistically in line with Avail's other code. We do make a concerted effort to keep things as uniform as possible. One thing to do is update the indents to tabs. You can do this in an editor with:
![image](https://github.com/AvailLang/Avail/assets/13374407/aefc7deb-b2c2-4177-8691-dc382f139c85)

followed by Option + Command + L (on Mac). You can also set the project Settings for Kotlin to use tabs just for the Avail Project:

![image](https://github.com/AvailLang/Avail/assets/13374407/6a55d9bd-05f0-474d-8a4b-e65d9ca41b08)


You can set the "Tab size" to 2 characters to match your style if you need to. Just be mindful of the 80 character preferred line size.

raulraja committed:
Apply Avail formatter style file and code review style comments

raulraja committed:
add return type

raulraja committed:
more style changes

raulraja committed:
change to tabs instead of space

raulraja committed:
apply visual guides and hardwrap 80

richATavail commented on avail-intellij-plugin/src/main/kotlin/org/availlang/intellij/plugin/module/AvailModuleWizardStep.kt:
```diff
@@ -31,24 +31,69 @@
 /**
  * The [ModuleWizardStep] used to set up a new Avail project.
  *
  * @author Richard Arriaga
```
Not gonna block this PR for this, but generally, if you work in a file someone else created, you should add your author tag below the other listed authors. So feel free to add your author tag here.

richATavail commented on avail/src/main/kotlin/avail/anvil/components/ComboBoxWithLabel.kt:
```diff
@@ -0,0 +1,137 @@
+ * A [JPanel] with a [GridBagLayout] that places a [JLabel] to the left of a
+ * [JComboBox].
+ *
+ * @param label The text for the [JLabel] to the left of the [JComboBox].
+ * @param items The items to be displayed in the [JComboBox].
+ * @param additionalComponents Additional components to be added to the bottom of the [JComboBox].
+ * @param emptySpaceRight Optional space to the right of the [JComboBox].
+ * @param emptySpaceLeft Optional space to the left of the [JLabel].
+ * @param panelBorder The border surrounding the [JPanel].
```
Just noticed these, go ahead and wrap the `@param` descriptions

```kotlin
/**
 * A [JPanel] with a [GridBagLayout] that places a [JLabel] to the left of a
 * [JComboBox].
 *
 * @param label 
 *   The text for the [JLabel] to the left of the [JComboBox].
 * @param items 
 *   The items to be displayed in the [JComboBox].
 * @param additionalComponents 
 *   Additional components to be added to the bottom of the [JComboBox].
 * @param emptySpaceRight 
 *   Optional space to the right of the [JComboBox].
 * @param emptySpaceLeft 
 *   Optional space to the left of the [JLabel].
 * @param panelBorder 
 *   The border surrounding the [JPanel].
 */
```

richATavail commented on avail/src/main/kotlin/avail/anvil/components/ComboBoxWithLabel.kt:
```diff
@@ -0,0 +1,137 @@
+					gridx = 0  // Align with the start of the combo box
+					gridy = index + 1  // Position under the combo box
+					gridwidth = 2  // Span across both label and combo box
+					insets = Insets(10, 0, 0, 0)  // Add some top margin for better spacing
```
This line runs beyond 80 characters. Bring the line comment above:

```kotlin
// Add some top margin for better spacing
insets = Insets(10, 0, 0, 0)
```

richATavail reviewed (changes requested):
Just two minor changes that I just now noticed. After that, you'll be good to merge! :)

raulraja committed:
PR review comments
- https://github.com/AvailLang/Avail/pull/310#discussion_r1364352069
- https://github.com/AvailLang/Avail/pull/310#discussion_r1364359836
- https://github.com/AvailLang/Avail/pull/310#discussion_r1364361056

richATavail reviewed (approved):
Looks good! Welcome to Avail!
